### PR TITLE
Adding unit tests

### PR DESCRIPTION
**Pull request by adishagarwal**

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -50,6 +53,11 @@ jobs:
 
       - name: 'npm lint'
         run: 'npm run lint'
+      
+      - uses: google-github-actions/auth@v2
```
Comment (sethvargo) [inline .github/workflows/test.yml:57]:
```suggestion
      - uses: 'google-github-actions/auth@v2'
```

File: package.json [hunk 2]
```diff
@@ -28,11 +28,13 @@
     "google-auth-library": "^9.4.2"
   },
   "devDependencies": {
+    "@types/sinon": "^17.0.1",
```
Comment (sethvargo) [inline package.json:31]: Please do not use `sinon` for mocking. Use the native Node.js mocks as we do elsewhere.

File: src/accessor.ts [hunk 3]
```diff
@@ -137,15 +132,19 @@ export class IACAccessor {
...
     private readonly baseURL: string,
     private readonly organizationId: string,
     private readonly scanTimeOut: number,
     private readonly scanStartTime: number,
     private readonly version: string,
+    private readonly httpClient?: HttpClient,
+    private readonly date?: Date,
   ) {
-    this.client = new HttpClient(USER_AGENT(version));
+    this.client = httpClient ?? new HttpClient(USER_AGENT(version));
```
Comment (sethvargo) [inline src/accessor.ts:147]: Why this change?

File: src/accessor.ts [hunk 4]
```diff
@@ -222,7 +217,7 @@ export class IACAccessor {
         throw new IACValidationException(
           /* statusCode= */ 500,
           `[Internal Error] Validation Service Endpoint Returned
-        invalid violations with one or more missing key Attributes, policyID : ${violation.policyId}, assetId : ${violation.assetId}`,
+        invalid violations with one or more missing key Attributes, policyID : ${violation.policyId ?? ''}, assetId : ${violation.assetId ?? ''}`,
```
Comment (sethvargo) [inline src/accessor.ts:220]: Do we expect these to be null?

File: test/accessor.test.ts [hunk 5]
```diff
@@ -0,0 +1,448 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { test } from 'node:test';
+import assert from 'node:assert';
+import { stub, SinonStub } from 'sinon';
```
Comment (sethvargo) [inline test/accessor.test.ts:19]: Please use `mock` from `node:test` instead

File: test/accessor.test.ts [hunk 6]
```diff
@@ -0,0 +1,448 @@
...
+  },
+  async (suite) => {
+    let httpClientStub: SinonStub;
+    let dateStub: SinonStub;
+    let accessor: IACAccessor;
+
+    suite.beforeEach(function () {
+      const httpClient = new HttpClient();
+      const date = new Date();
+      httpClientStub = stub(httpClient, 'request');
```
Comment (sethvargo) [inline test/accessor.test.ts:63]: There are examples of how to mock this node:test in deploy-cloud-functions.

**Review by sethvargo — COMMENTED** [review]
I think the test files traditionally live under the `tests/` directory (not `test/`)

File: src/accessor.ts (hunk 3 above)
Comment (adishagarwal) [reply src/accessor.ts:147]: We want to mock the http client in our UT's. In order to do that, i have made http client an optional param to the constructor, so that the mocked http client could be passed from the UT's.

File: src/accessor.ts (hunk 4 above)
Comment (adishagarwal) [reply src/accessor.ts:220]: Ideally no, we have place this check to ensure that these fields are non null in the api response at runtime.

File: package.json (hunk 2 above)
Comment (adishagarwal) [reply package.json:31]: Filed an issue on the repo to migrate from sinon: https://github.com/google-github-actions/analyze-code-security-scc/issues/15

File: test/accessor.test.ts (hunk 5 above)
Comment (adishagarwal) [reply test/accessor.test.ts:19]: Filed an issue on the repo to migrate from sinon: https://github.com/google-github-actions/analyze-code-security-scc/issues/15

File: test/accessor.test.ts (hunk 6 above)
Comment (adishagarwal) [reply test/accessor.test.ts:63]: Ack.

File: test/accessor.test.ts [hunk 7]
```diff
@@ -0,0 +1,448 @@
...
+      dateStub = stub(date, 'getTime');
+      dateStub.returns(0);
+      accessor = new IACAccessor(
+        BASE_URL,
+        ORGANIZATION_ID,
+        /* scanTimeOut= */ 60000,
```
[the remark this answers is not in the document]
Comment (adishagarwal) [reply test/accessor.test.ts:69]: These are the param's name, added these to have better context for reader's.

File: .github/workflows/test.yml (hunk 1 above)
Comment (adishagarwal) [reply .github/workflows/test.yml]: Done.

File: test/utils.test.ts [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+        error: 'failure_criteria validation failed : multiple severities of type LOW found',
+      },
+      {
+        name: 'no severities',
+        input: 'Operator: OR',
+        error: 'failure_criteria validation failed : no severity mentioned',
+      },
+    ];
+
+    cases.forEach((tc) => {
```
[the remark this answers is not in the document]
Comment (adishagarwal) [reply test/utils.test.ts]: Done.

File: test/utils.test.ts [hunk 9]
```diff
@@ -0,0 +1,341 @@
...
+              assetId: 'asset-id',
+              policyId: 'policy-id',
+              severity: Severity.High,
+            },
+          ],
+        },
+        expected: true,
+      },
+    ];
+    cases.forEach((tc) => {
```
[the remark this answers is not in the document]
Comment (adishagarwal) [reply test/utils.test.ts]: Done.

File: src/accessor.ts (hunk 3 above)
Comment (sethvargo) [reply src/accessor.ts:147]: You can pretty easily mock `HttpClient.prototype` without needing the dependency injection.

File: test/accessor.test.ts (hunk 5 above)
Comment (sethvargo) [reply test/accessor.test.ts]: I know your launch is soon, but I'd prefer doing it in this PR. It will also resolve the unnecessary dependency injection of a custom httpClient, since you can mock the prototype directly with mock from node:test.

File: test/accessor.test.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (sethvargo) [reply test/accessor.test.ts]: My IDE automatically does that. I don't think we want this inline in the code.

[the remark this answers is not in the document]
Comment (sanyam803) [reply test/accessor.test.ts]: Having these annotations is inline with typescript readability.

File: src/accessor.ts (hunk 3 above)
Comment (sanyam803) [reply src/accessor.ts:147]: I believe this is inline with testing best practices that we inject any dependency so that it's unit testable.

Comment (adishagarwal) [reply src/accessor.ts]: Removed the dependency, we have also removed sinon from UT's now.

File: test/accessor.test.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (adishagarwal) [reply test/accessor.test.ts]: Removed these.

**Review by sanyam803 — APPROVED** [review]

**Review by verbanicm — APPROVED** [review]
